docroot/sites/all/modules/contrib/user_cancel_immediate/user_cancel_immediate.module

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/**
+ * @file
+ * Module file for user_cancel_immediate
+ * Change user cancellation to be immmediate, without any email confirmation.
+ *
+ * Thanks to Clive, http://drupal.stackexchange.com/a/32427/449
+ */
+
+
+/**
+ * Implements hook_form_FORMID_alter()
+ * Adds a custom submit handler
+ */
+function user_cancel_immediate_form_user_cancel_confirm_form_alter(&$form, &$form_state, $form_id) {
+  $form['#submit'][] = '_user_cancel_immediate_user_cancel_form_submit';
+}
+
+/**
+ * Additional submit handler for user cancel operation - cancel account instead of
+ * sending email and having them click the link.
+ * Suggested by http://drupal.stackexchange.com/a/32427/449
+ *
+ * @param $form
+ * @param $form_state
+ */
+function _user_cancel_immediate_user_cancel_form_submit(&$form, &$form_state) {
+  $uid = $form['uid']['#value'];
+
+  if (!user_access('administer users') && $uid == $GLOBALS['user']->uid) {
```

Comment:
Would this not run into trouble if the 'user cancel own account' permission is enabled for ordinary users?